File: cpp/graph/bfs.cc
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

class Node {
 public:
  Node(int i_) : i(i_) {}

  void AddAdj(int u) {
    adj.push_back(u);
  }
  
  int i;
  vector<int> adj;
};

int BFS(const vector<Node> &graph, int start, int find) {
  queue<int> q;
  unordered_set<int> seen;
  q.push(start);
  seen.insert(start);
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    if (graph[u].i == find) return u;
    for (auto &v : graph[u].adj) {
      if (seen.find(v) == seen.end()) {
        seen.insert(v);
        q.push(v);
      }
    }
  }
  return -1;
}

File: python/graph/bfs.py
from collections import deque

class Node:
    def __init__(self, i):
        self.i = i
        self.adj = []

    def add_adj(self, a):
        self.adj.append(a)


def BFS(start, find):
    q = deque([start])
    seen = set([start])
    while q:
        u = q.popleft()
        if u.i == find:
            return u
        for v in u.adj:
            if v.i not in seen:
                seen.add(v.i)
                q.append(v)
    return None

